packages/notifications/push/amplify_push_notifications/ios/Classes/PushNotificationEventsStreamHandler.swift

```diff
@@ -0,0 +1,141 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import Foundation
+import Flutter
+
+private let channelNamePrefix = "com.amazonaws.amplify/push_notification/"
+
+enum NativeEvent {
+    case tokenReceived
+    case notificationOpened
+    case launchNotificationOpened
+    case foregroundMessageReceived
+    case backgroundMessageReceived
+
+    var eventName: String {
+        switch self {
+        case .tokenReceived:
+            return "TOKEN_RECEIVED"
+        case .notificationOpened:
+            return "NOTIFICATION_OPENED"
+        case .launchNotificationOpened:
+            return "LAUNCH_NOTIFICATION_OPENED"
+        case .foregroundMessageReceived:
+            return "FOREGROUND_MESSAGE_RECEIVED"
+        case .backgroundMessageReceived:
+            return "BACKGROUND_MESSAGE_RECEIVED"
+        }
+    }
+
+    var eventChannelName: String {
+        return "\(channelNamePrefix)event/\(self.eventName)"
+    }
+}
+
+let eventChannels = [
+    NativeEvent.tokenReceived.eventChannelName,
+    NativeEvent.notificationOpened.eventChannelName,
+    NativeEvent.launchNotificationOpened.eventChannelName,
+    NativeEvent.foregroundMessageReceived.eventChannelName,
+    NativeEvent.backgroundMessageReceived.eventChannelName
+]
+
+struct AmplifyPushNotificationsEvent {
+    var event: NativeEvent
+    var payload: [AnyHashable: Any]
+
+    func toMap() -> [AnyHashable: Any] {
+        return [
+            "eventType": event.eventName,
+            "payload": payload
+        ]
+    }
+}
+
+class PushNotificationEventsStreamHandler: NSObject, FlutterStreamHandler {
+    private let eventChannel: FlutterEventChannel
+    private let eventType: NativeEvent
+    private var eventQueue: [AmplifyPushNotificationsEvent] = []
+    private var eventSink: FlutterEventSink?
+
+    init(eventType: NativeEvent, binaryMessenger: FlutterBinaryMessenger) {
+        self.eventType = eventType
+        eventChannel = FlutterEventChannel(name: eventType.eventChannelName, binaryMessenger: binaryMessenger)
+        super.init()
+        eventChannel.setStreamHandler(self)
+    }
+
+    func onListen(
+        withArguments arguments: Any?,
+        eventSink events: @escaping FlutterEventSink
+    ) -> FlutterError? {
+        eventSink = events
+        flushEvents()
+        return nil
+    }
+
+    func onCancel(
+        withArguments arguments: Any?
+    ) -> FlutterError? {
+        eventSink = nil
+        eventQueue.removeAll()
+        return nil
+    }
+
+    func sendEvent(payload: [AnyHashable: Any]) {
+        let event = AmplifyPushNotificationsEvent(event: eventType, payload: payload)
+        if let eventSink = eventSink {
+            eventSink(event.toMap())
+        } else {
+            eventQueue.append(event)
+        }
+    }
+
+    func sendError(error: FlutterError) {
+        if let eventSink = eventSink {
+            eventSink(error)
```

Comment:
This should go to the `eventQueue` too. I ran into this problem where there may be an error before the eventsink is ready, that time the error gets lost. Going through the queue will ensure error gets emitted when the stream is listened to and sink is ready.

Reply:
updated to queue the error as well.